presto-spi/src/main/java/com/facebook/presto/spi/block/SliceArrayBlock.java

```diff
@@ -45,7 +50,7 @@ public SliceArrayBlock(int positionCount, Slice[] values)
         this.values = values;
 
         sizeInBytes = getSliceArraySizeInBytes(values, 0, values.length);
-        retainedSizeInBytes = INSTANCE_SIZE + getSliceArrayRetainedSizeInBytes(values);
+        retainedSizeInBytes = INSTANCE_SIZE + getSliceArrayRetainedSizeInBytes(values, valueSlicesAreDistinct);
```

Comment:
Add comment that gives an intuitive high-level explanation of why this optimization is important.